### PR TITLE
[NVHPC] add a possibility to control default CUDA version

### PR DESCRIPTION
**Pull request by toxa81**

### CODE REVIEW

**Review by tldahlgren — DISMISSED** [review]
LGTM. Defer to merge to give @samcmilll time to comment.

**Comment by samcmill** [comment]
LGTM.  But please realize that this option only applies if a CUDA driver cannot be detected.  If a driver is detected, then the HPC SDK selects the best match from the CUDA versions included with the HPC SDK.

**Comment by msimberg** [comment]
I don't have anything fundamentally against this so happy to have it merged if it's useful. The concern by @samcmill is a bit nasty because it may mean that an installation with `cuda_version=X` may in fact be using `Y` and it's context dependent...

Another concern is that there are only three CUDA versions per NVHPC install and currently `cuda_version` allows any version (any string really). I don't know if it's worth adding conflicts for that. Will the install fail if it's set to a nonexistent version or worse to a random string?

Finally, @samcmill would exporting `NVHPC_CUDA_HOME` in dependent packages' environments enforce a particular version or is that orthogonal?

Finally, I guess it's currently not possible to install `nvhpc` _only for the compilers_ and make it depend on the `cuda` package for libraries? If that were possible, it'd make things much more explicit and allow choosing an exact version of CUDA with the regular `^cuda@X.Y.Z` syntax...

**Comment by samcmill** [comment]
> would exporting `NVHPC_CUDA_HOME` in dependent packages' environments enforce a particular version or is that orthogonal?

Yes, that would enforce a particular version.  However, `NVHPC_CUDA_HOME` is a big hammer with lots of rough edges.  I do not recommend it for general use.

The documented way to specify a specific one of the bundled CUDA versions is to use the `-gpu=cudaX.Y` compiler option.

**Review by msimberg — DISMISSED** [review]
Ok for me as the new variant doesn't change existing behaviour (default is to not set `NVHPC_DEFAULT_CUDA`).

File: var/spack/repos/builtin/packages/nvhpc/package.py
Comment (msimberg) [inline var/spack/repos/builtin/packages/nvhpc/package.py]: Not a blocking comment, but should this rather be called:
```suggestion
    variant("default_cuda", default="default", description="Default CUDA version, for example 11.8")
```
since it doesn't actually force a particular version?

**Comment by msimberg** [comment]
Pinging @ax3l and @Rombur as well for comments as maintainers of the `cuda` package.

File: var/spack/repos/builtin/packages/nvhpc/package.py
Comment (Rombur) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: I also like `default_version` better.

Comment (toxa81) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: resolved

**Review by Rombur — APPROVED** [review]

**Review by msimberg — DISMISSED** [review]
@toxa81 you'll need to fix the formatting as reported by https://github.com/spack/spack/actions/runs/5974558155/job/16208952590?pr=38909#step:6:19.

Looks good to me otherwise.

**Comment by msimberg** [comment]
@spackbot fix style

**Review by msimberg — APPROVED** [review]

**Pull request merged**
